dre-service/src/main/java/io/cdap/re/RulesDB.java

```diff
@@ -147,376 +199,399 @@ public void updateRule(String id, RuleRequest rule) throws RuleNotFoundException
     List<String> thens = new ArrayList<>();
-    for (String t : then.split(";")) {
-      thens.add(t);
-    }
-    result.put(Bytes.toString(ACTION), thens);
+    Collections.addAll(thens, Objects.requireNonNull(then).split(";"));
+    rule.put(ACTION_COL, thens);
 
-    return result;
+    return rule;
   }
 
-  public void deleteRule(String id) throws RuleNotFoundException {
-    Row row = rules.get(toKey(id));
-    if (row.isEmpty()) {
+  void deleteRule(String id) throws RuleNotFoundException, IOException {
+    Collection<Field<?>> keyFields = new ArrayList<>();
+    keyFields.add(Fields.stringField(ID_COL, id));
+
+    Optional<StructuredRow> optionalStructuredRow = rulesTable.read(keyFields);
+
+    if (!optionalStructuredRow.isPresent()) {
       throw new RuleNotFoundException(
         String.format("Rule '%s' does not exist. Create it first with a POST request.", id)
       );
     }
-    rules.delete(toKey(id));
+
+    rulesTable.delete(keyFields);
   }
 
-  public String retrieveUsingRuleTemplate(String id) throws RuleNotFoundException {
-    Row row = rules.get(toKey(id));
-    if (row.isEmpty()) {
+  String retrieveUsingRuleTemplate(String id) throws RuleNotFoundException, IOException {
+    Collection<Field<?>> keyFields = new ArrayList<>();
+    keyFields.add(Fields.stringField(ID_COL, id));
+
+    Optional<StructuredRow> optionalStructuredRow = rulesTable.read(keyFields);
+
+    if (!optionalStructuredRow.isPresent()) {
       throw new RuleNotFoundException(
         String.format("Rule '%s' does not exist. Create it first with a POST request.", id)
       );
     }
 
-    return String.format(RULE_TEMPLATE, row.getString(ID), row.getString(DESCRIPTION),
-                                row.getString(CONDITION), row.getString(ACTION));
+    StructuredRow row = optionalStructuredRow.get();
+
+    return String.format(RULE_TEMPLATE, row.getString(ID_COL), row.getString(DESCRIPTION_COL),
+                         row.getString(CONDITION_COL), row.getString(ACTION_COL));
   }
 
-  public void createRulebook(RulebookRequest rb) throws RulebookAlreadyExistsException, RuleNotFoundException {
-    Row row = rulebook.get(toKey(rb.getId()));
-    if (!row.isEmpty()) {
+  void createRulebook(RulebookRequest rulebookRequest)
+    throws RulebookAlreadyExistsException, RuleNotFoundException, IOException {
+    Collection<Field<?>> rulebookKeyFields = new ArrayList<>();
+    rulebookKeyFields.add(Fields.stringField(ID_COL, rulebookRequest.getId()));
+
+    Optional<StructuredRow> optionalRulebookStructuredRow = rulebookTable.read(rulebookKeyFields);
+
+    if (optionalRulebookStructuredRow.isPresent()) {
       throw new RulebookAlreadyExistsException(
-        String.format("Rulebook '%s' already exists. Either update it or delete it and create it.", rb.getId())
+        String
+          .format("Rulebook '%s' already exists. Either update it or delete it and create it.", rulebookRequest.getId())
```

Comment:
nit: it's more readable if you put the rulebookRequest.getId() part on this line and have everything else in the line above.

Reply:
Done